File: Python3/486. Predict the Winner.py
class Solution:
    def PredictTheWinner(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """        
        length = len(nums) # 独立完成了一次dp问题
        if length == 1:
            return True
        dp = {}
        def recursion(start, end):
            if end == start + 1:
                if nums[start] > nums[end]:
                    dp[(start, end)] = (nums[start], nums[end])
                else:
                    dp[(start, end)] = (nums[end], nums[start])
                return dp[(start, end)]
            if (start, end) in dp:
                return dp[(start, end)]
            min1, max1 = recursion(start+1, end)
            max1 += nums[start]
            min2, max2 = recursion(start, end-1)
            max2 += nums[end]
            if max1 > max2:
                dp[(start, end)] = (max1, min1)
            else:
                dp[(start, end)] = (max2, min2)
            return dp[(start, end)]
        a, b = recursion(0, length-1)
        return a >= b

File: Golang/486. Predict the Winner.go
func PredictTheWinner(nums []int) bool {
    length := len(nums)
    if length == 1 {
        return true
    }
    dp := make([][]int, 2000)
    array := recursion(&dp, &nums, 0, length-1)
    return array[0] >= array[1]
}

func recursion(dp *[][]int, nums *[]int, start, end int) []int {
    if len((*dp)[start*100+end]) > 0 {
        return (*dp)[start*100+end]
    }
    if end == (start + 1) {
        if (*nums)[start] > (*nums)[end] {
            (*dp)[start*100+end] = []int{(*nums)[start], (*nums)[end]}
        } else {
            (*dp)[start*100+end] = []int{(*nums)[end], (*nums)[start]}
        }
        return (*dp)[start*100+end]
    }
    array1 := recursion(dp, nums, start+1, end)
    array1[1] += (*nums)[start]
    array2 := recursion(dp, nums, start, end-1)
    array2[1] += (*nums)[end]
    if array1[1] > array2[1] {
        (*dp)[start*100+end] = []int{array1[1], array1[0]}
    } else {
        (*dp)[start*100+end] = []int{array2[1], array2[0]}
    }
    return (*dp)[start*100+end]
}
